### PR TITLE
test/throttle: kill tests exercising dtor of Throttle classes

### PR DESCRIPTION
**Pull request by tchaikov**
quote from Adam,

> let's just remove all the destructor tests. Since it's just exercising
> an assertion, it's most likely to be complexity that doesn't buy us
> much.

Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

File: src/test/common/Throttle.cc [hunk 1]
```diff
@@ -357,15 +357,15 @@ TEST(BackoffThrottle, destruct) {
...
-	  });
-      }
+      auto& t = *throttle;
+      std::thread runner([&t]() {
+	  usleep(5);
+	  t.get(6);
+	});
       // No equivalent of get_or_fail()
       std::this_thread::sleep_for(std::chrono::milliseconds(250));
+      throttle.reset();
```
Comment (kungf) [inline src/test/common/Throttle.cc]: BackoffThrottle doesn't not have reset() member function.

File: src/test/common/Throttle.cc [hunk 2]
```diff
@@ -427,16 +427,16 @@ TEST(OrderedThrottle, destruct) {
...
-      }
+      throttle->start_op(new FunctionContext(NULL));
+      auto& t = *throttle;
+      std::thread runner([&t]() {
+	  usleep(5);
+	  t.start_op(new FunctionContext(NULL));
+	});
       // No equivalent of get_or_fail()
       std::this_thread::sleep_for(std::chrono::milliseconds(250));
+      throttle.reset();
```
Comment (kungf) [inline src/test/common/Throttle.cc]: OrderedThrottle  does not have reset() member function.

**Comment by kungf** [comment]
@adamemerson what't do you want to test in TEST(OrderedThrottle, destruct) or TEST(BackoffThrottle, destruct), to test the assert(waiters.empty()) in their dtor? if so, thread may be not need joined.

File: src/test/common/Throttle.cc (hunk 1 above)
Comment (tchaikov) [reply src/test/common/Throttle.cc]: `unique_ptr` does.

Comment (kungf) [reply src/test/common/Throttle.cc]: learned

**Comment by adamemerson** [comment]
Jenkins, retest this please.

**Comment by adamemerson** [comment]
The main thing we want to test for is that the program dies with an assert or terminate or something if someone tries to destroy the throttle with outstanding waiters on the condition variable, since destroying a condition variable with outstanding waiters is undefined and in newer versions of glibc it just makes your process hang forever.

**Comment by adamemerson** [comment]
I'm not sure if we need to worry about exercising the destructors that much. That they assert if you destroy a throttle with waiters isn't really different from other asserts we have. The only reason I made a test for it is because there was already a destruct test that assumed the (undefined) behavior was to wake the waiters on destruction. It probably would have been just as good to slap the asserts in and delete the test.

**Comment by tchaikov** [comment]
@adamemerson i am good either way: to remove the test or merge this PR.

**Comment by tchaikov** [comment]
@adamemerson updated and repushed.

**Pull request merged**